service/gcs/main.go

```diff
@@ -5,21 +5,60 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 	"syscall"
 	"time"
 
+	"github.com/Microsoft/opengcs/internal/cgrouputils"
 	"github.com/Microsoft/opengcs/internal/oc"
 	"github.com/Microsoft/opengcs/internal/runtime/hcsv2"
 	"github.com/Microsoft/opengcs/service/gcs/bridge"
 	"github.com/Microsoft/opengcs/service/gcs/core/gcs"
 	"github.com/Microsoft/opengcs/service/gcs/runtime/runc"
 	"github.com/Microsoft/opengcs/service/gcs/transport"
 	"github.com/containerd/cgroups"
+	v1 "github.com/containerd/cgroups/stats/v1"
 	oci "github.com/opencontainers/runtime-spec/specs-go"
 	"github.com/sirupsen/logrus"
 	"go.opencensus.io/trace"
 )
 
+func memoryLogFormat(metrics *v1.Metrics) logrus.Fields {
+	return logrus.Fields{
+		"memory-usage":       metrics.Memory.Usage.Usage,
+		"memory-usage-max":   metrics.Memory.Usage.Max,
+		"memory-usage-limit": metrics.Memory.Usage.Limit,
+		"swap-usage":         metrics.Memory.Swap.Usage,
+		"swap-usage-max":     metrics.Memory.Swap.Max,
+		"swap-usage-limit":   metrics.Memory.Swap.Limit,
+		"kernel-usage":       metrics.Memory.Kernel.Usage,
+		"kernel-usage-max":   metrics.Memory.Kernel.Max,
+		"kernel-usage-limit": metrics.Memory.Kernel.Limit,
+	}
+}
+
+func readMemoryEvents(efdFile *os.File, cgName string, threshold int64, cg cgroups.Cgroup) {
+	// Buffer must be >= 8 bytes for eventfd reads
+	// http://man7.org/linux/man-pages/man2/eventfd.2.html
+	buf := make([]byte, 8)
+	for {
+		_, err := efdFile.Read(buf)
+		if err != nil {
+			logrus.WithError(err).Warn("opengcs::readEvents - failed to read from eventfd")
+			return
+		}
+		msg := fmt.Sprintf("opengcs::readMemoryEvents - memory usage for cgroup '%s' exceeded '%d' byte threshold", cgName, threshold)
```

Comment:
Can we make this msg not dynamic and just incorporate the information into fields?

Reply:
(That way it's easier to parse it later.)

Reply:
The cgName and threshold can be passed into memoryLogFormat and moved to fields yes. The dynamic aspect of the message  carries through to the lines below this if we fail to retrieve the stats as we don't have anything to log besides the name and threshold though. So I guess there's two questions regarding if we fail to retrieve the stats 
1. Should I keep the appending of the "- failed to gather memory metrics"
2. If we fail to gather the metrics should we just log the message with cgName and threshold as fields or should we stay consistent with a success and log all of the fields but everything besides cgName and threshold be default values.

Reply:
@kevpar can you provide guidance

Reply:
You can pre-populate a `logrus.Entry` with the initial fields:

```
msg := "monitored cgroup memory exceeded threshold"
entry := logrus.WithFields(logrus.Fields{
    "cgroup": cgName,
    "thresholdBytes": threshold,
})
time.Sleep(time.Second)
metrics, err := cg.Stat()
if err != nil {
    entry.WithError(err).Error(msg)
} else {
    logrus.WithFields(memoryLogFormat(metrics)).Warn(msg)
}
```

Also:

- If we fail to get the memory stats we should log that as an error, rather than a warning.
- We should log the error that we receive from `cg.Stat()`.
- I'm not sure we need to add to the message in the failure case, we can probably just log it at a different level and include the error as a field.